circularqueue_test.go

```diff
@@ -0,0 +1,249 @@
+			expectedPeek: [][]circularInt{{6, 7, 8, 9}, {7, 8, 9, 10, 11, 12}},
+			add:          [][]circularInt{{8, 9}, {10, 11, 12}},
+			dequeueTimes: []int{5, 1},
+		},
+		{
+			starting:     []circularInt{1, 2, 3, 4, 5, 6, 7},
+			name:         "remove then add #3",
+			expected:     []circularInt{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12},
+			expectedPeek: [][]circularInt{{6, 7, 8}, {}, {13}},
+			add:          [][]circularInt{{8}, {9, 10, 11}, {12, 13}},
+			dequeueTimes: []int{5, 7, 1},
+		},
+		{
+			starting:     []circularInt{1, 2, 3, 4, 5, 6, 7},
+			name:         "remove then add #4",
+			expected:     []circularInt{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11},
+			expectedPeek: [][]circularInt{{6, 7, 8}, {11}, {12}},
+			add:          [][]circularInt{{8}, {9, 10, 11}, {12}},
+			dequeueTimes: []int{5, 5, 1},
+		},
+		{
+			starting:     []circularInt{1, 2, 3, 4, 5, 6, 7},
+			name:         "remove too many too early then add one in",
+			expected:     []circularInt{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11},
+			expectedPeek: [][]circularInt{{6, 7, 8}, {}, {12}},
+			add:          [][]circularInt{{8}, {9, 10, 11}, {12}},
+			dequeueTimes: []int{5, 33, 0},
+		},
+		{
+			starting:     []circularInt{1, 2, 3, 4, 5, 6, 7},
+			name:         "remove everyone",
+			expected:     []circularInt{1, 2, 3, 4, 5, 6, 7},
+			expectedPeek: [][]circularInt{{6, 7}, {}},
+			add:          [][]circularInt{{}, {}},
+			dequeueTimes: []int{5, 3},
+		},
+		{
+			starting:     []circularInt{1, 2, 3, 4, 5, 6, 7},
+			name:         "remove everything but one then fill it up again",
+			expected:     []circularInt{1, 2, 3, 4, 5, 6, 7},
+			expectedPeek: [][]circularInt{{6, 7}, {}, {8, 9, 10, 11}},
+			add:          [][]circularInt{{}, {}, {8, 9, 10, 11}},
+			dequeueTimes: []int{5, 3, 0},
+		},
+	}
+	for _, c := range cases {
+		t.Run(c.name, func(t *testing.T) {
+			q := newCircularIntCircularQueue(c.starting...)
+			res := []circularInt{}
+			for j := 0; j < len(c.dequeueTimes); j++ {
+				for i := range c.add[j] {
+					q.Enqueue(c.add[j][i])
+				}
+				peekRes := []circularInt{}
+				if len(peekRes) > 0 {
+					for i := 0; i < q.Len(); i++ {
+						peekRes = append(peekRes, q.Peek(i))
+					}
+					if !reflect.DeepEqual(peekRes, c.expectedPrePeek[j]) {
+						t.Errorf("expected peeking before we called next, on step %d we would see %v, got %v", j, c.expectedPeek[j], peekRes)
+					}
+				}
+				for i := 0; i < c.dequeueTimes[j]; i++ {
+					if q.Len() > 0 {
+						res = append(res, q.Peek(0))
+						q.Dequeue(1)
+					}
+				}
+				peekRes = []circularInt{}
+				for i := 0; i < q.Len(); i++ {
+					peekRes = append(peekRes, q.Peek(i))
+				}
+				if !reflect.DeepEqual(peekRes, c.expectedPeek[j]) {
+					t.Errorf("expected peeking, on step %d we would see %v, got %v", j, c.expectedPeek[j], peekRes)
+				}
+			}
+			if !reflect.DeepEqual(res, c.expected) {
+				t.Errorf("expected %v, got %v", c.expected, res)
+			}
+		})
+
+	}
+}
+
+func Test_leakCircularBuf(t *testing.T) {
```

Comment:
It might be easier to use Benchmarks for this sort of test, it gives you the number of allocations. (for the future, this is fine as-is)

Reply:
This isn't for testing allocations, but for testing to make sure things get properly de-allocated (a test against leaks).